libs/core/synchronization/include/hpx/synchronization/spinlock.hpp

```diff
@@ -49,7 +49,30 @@ namespace hpx { namespace lcos { namespace local {
         {
             HPX_ITT_SYNC_PREPARE(this);
 
-            while (!acquire_lock())
+            // Checking for the value in is_locked() ensures that
+            // acquire_lock is only called when is_locked computes
+            // to false. This way we spin only on a load operation
+            // which minimizes false sharing that comes with an
+            // exchange operation.
+            // Consider the following cases:
+            // 1. Only one thread wants access critical section:
+            //      is_locked() -> false; computes acquire_lock()
+            //      acquire_lock() -> false (new value set to true)
+            //      Thread acquires the lock and moves to critical
+            //      section.
+            // 2. Two threads simultaneously access critical section:
+            //      Thread 1: is_locked() || acquire_lock() -> false
+            //      Thread 1 acquires the lock and moves to critical
+            //      section.
+            //      Thread 2: is_locked() -> true; execution enters
+            //      inside while without computing acquire_lock().
+            //      Thread 2 yields while is_locked() computes to
+            //      false. Then it retries doing is_locked() -> false
+            //      followed by an acquire_lock() operation.
+            //      The above order can be changed arbitrarily but
+            //      the nature of execution will still remain the
+            //      same.
+            while (is_locked() || !acquire_lock())
             {
                 util::yield_while([this] { return is_locked(); },
                     "hpx::lcos::local::spinlock::lock");
```

Comment:
Would this be equivalent, and possibly simpler?

```suggestion
            do 
            {
                util::yield_while([this] { return is_locked(); },
                    "hpx::lcos::local::spinlock::lock");
            } while (!acquire_lock())
```
(ignore the mismatched lines, github doesn't let me comment on lines below the diff).

Can we construct a small test to verify the effect of this? If nothing else, running the `future_overhead` benchmark, which relies on (mostly uncontended) spinlocks.

Thanks for looking into this!

Reply:
Good catch! We can certainly do it this way to make it look simpler. This will also ensure that there is a single load operation when leaving the loop (instead of the current 2). I'll correct it right away.

Reply:
@msimberg Let me write a small high contention test to test the performance. I do not suspect any significant speedup but decreasing the CPU bus traffic may potentially help us with performance somewhere else.

Reply:
This change will take effect in the uncontended case mostly. With higher contention false sharing is still an issue. But this can't be resolved that easily. Thanks for looking into this.

Would you mind fixing the `util::spinlock` in a similar way? This needs to be changed to try a load first as well: https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/thread_support/include/hpx/thread_support/spinlock.hpp#L48.

Reply:
@hkaiser the changes in the commit also include changes to the link you directed to, or am I getting something wrong?

Reply:
Perfect, I didn't see that. Thanks!